HAWQ-1480 - Added feature for packing a core file

outofmem0ry opened the pull request:
Currently there is no way to packing a core file with its context – executable, application and system shared libraries in hawq. This information can be later unpacked on another system and helps in debugging. It is a useful feature to quickly gather all the data needed from a crash/core generated on the system to analyze it later.
Another open source project, greenplum, uses a script https://github.com/greenplum-db/gpdb/blob/master/gpMgmt/sbin/packcore to collect this information. Merging this into hawq.

linwen commented on tools/sbin/packcore:
```diff
@@ -0,0 +1,262 @@
+# specific language governing permissions and limitations
+# under the License.
+
+# Copyright Pivotal 2014
```
I don't suggest add "Copyright Pivotal 2014" here, since hawq is under Apache license.

outofmem0ry replied:
@linwen It was in the original code repo I merged it from. Wanted to keep it there for reference/credits. Something similar to what is done here- incubator-hawq/tools/sbin/hawqstandbywatch.py

Let me know if it is feasible removing copyright from merged code. I can update my branch.

linwen replied:
I think it is better to remove it, also the ""Copyright Pivotal" in hawqstandbywatch.py. How do you think of this Roman? @rvs Thanks!

vVineet replied:
We need to remove Pivotal copyright since we are contributing this code in Apache and doing so under Apache license.

edespino commented on tools/sbin/packcore:
```diff
@@ -0,0 +1,262 @@
+
+def _getLibraryListWithLDD(binary):
+    # We manually seed this with a few libraries that are missed
+    # This may not be needed for all proces, but will round out the
```
"proces" is misspelled.

edespino commented on tools/sbin/packcore:
```diff
@@ -0,0 +1,262 @@
+    u = '''%prog [options] core_file
+This will create an archive with the core file and all required
+libraries for analysis.  The preference is to use GDB so that we can
+resolve dependancies for extensions.'''
```
"dependancies" is misspelled.

edespino commented:
Aside from the two misspellings identified and the pending determination to keep or remove the "Copyright Pivotal 2014" string, this PR looks good.  I have run the packcore against a HAWQ instance and it performed as expected.  Will we need to add documentation for this utility?

edespino commented:
Additionally, is a test being added to validate the functionality of packcore?

linwen commented:
Ed, I agree with you on adding documentation for this utility. I have some concern on adding a test case, is it a little bit strange to have a core dump file in source repository(maybe two, OSX and Linux)? Or write a test case which can trigger a core dump, then run this script.

outofmem0ry commented:
@edespino @vVineet @linwen - Thank you for the reviews and comments. I have incorporated the changes and removed the copyright for now.

@linwen What kind of documentation do you have in mind? Here are some options -
- A separate markdown explaining the use case and how to, committed to this repository
- Use case and how to use packcore, committed under - incubator-hawq-docs/markdown/admin/

Additional suggestions are welcomed.

edespino commented:
@outofmem0ry - Thanks for addressing the issues. FYI: It is helpful to push the changes to your repo:branch (outofmem0ry:feature-packcore) that this PR is based on so we can see that the issues were indeed addressed. The PR will automatically get updated when you push to your dev branch. Thanks, -=e

edespino commented:
@outofmem0ry - It might be helpful to check with David Yozie (@dyozie) to see if we can get this into the public HAWQ user/admin documentation.

dyozie commented:
I doesn't look like Greenplum documents packcore either.

edespino commented:
I'm thinking from the perspective of an Apache developer. They shouldn't have to go to external corporate documentation sets or knowledge bases (https://discuss.pivotal.io/hc/en-us/articles/201081408-How-to-collect-core-files-for-analysis) to get useful information about a utility (packcore) included with the HAWQ distribution. As it stands, the utility (packcore) will be automatically bundled with all future HAWQ releases.

outofmem0ry commented:
@dyozie correct, it is not documented in greenplum.
@edespino - Totally agree with your point. Keeping that in mind, earlier, I suggested documenting it under incubator-hawq-docs/markdown/admin/ as this tool is more in an administrative capacity. 

Let me know if that sounds good and I can start with the documentation.

edespino commented:
@dyozie - is it easy for @outofmem0ry to add docs for this utility (packcore) in incubator-hawq-docs/markdown/admin/. Can you provide some guidance? It will be helpful to share this with the dev community for future reference.

dyozie commented:
Yes - it should be straightforward to add this information, and docs can
provide further assistance via the PR.

-David

On Thu, Jun 8, 2017 at 9:22 AM, Ed Espino <notifications@github.com> wrote:

> @dyozie <https://github.com/dyozie> - is it easy for @outofmem0ry
> <https://github.com/outofmem0ry> to add docs for this utility (packcore)
> in incubator-hawq-docs/markdown/admin/. Can you provide some guidance? It
> will be helpful to share this with the dev community for future reference.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-hawq/pull/1251#issuecomment-307154762>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFbpVqKXGUwWZePsUGLPQFv8P9plQvyWks5sCB_IgaJpZM4Nw2Wo>
> .
>

outofmem0ry commented:
@dyozie @edespino - submitted documentation under [incubator-hawq-docs pull Request #123](https://github.com/apache/incubator-hawq-docs/pull/123).

linwen commented:
LGTM

edespino commented:
This PR (sans documentation which is included in PR #123) LGTM.

linwen commented:
merged, this pr can be closed now. Thanks!

linwen commented:
Shubham, this PR has been merged into master. Would you please close it? Thanks!

outofmem0ry commented:
Thank you for the merge @linwen. Marking it closed.
